tracer/src/Datadog.Trace/RuntimeMetrics/MemoryMappedCounters.cs

```diff
@@ -0,0 +1,593 @@
+// <copyright file="MemoryMappedCounters.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#if NETFRAMEWORK
+
+using System;
+using System.IO.MemoryMappedFiles;
+using System.Runtime.InteropServices;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Util;
+using Datadog.Trace.Vendors.StatsdClient;
+
+namespace Datadog.Trace.RuntimeMetrics
+{
+    internal class MemoryMappedCounters : IRuntimeMetricsListener
+    {
+        private const string GarbageCollectionMetrics = $"{MetricsNames.Gen0HeapSize}, {MetricsNames.Gen1HeapSize}, {MetricsNames.Gen2HeapSize}, {MetricsNames.LohSize}, {MetricsNames.ContentionCount}, {MetricsNames.Gen0CollectionsCount}, {MetricsNames.Gen1CollectionsCount}, {MetricsNames.Gen2CollectionsCount}";
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<PerformanceCountersListener>();
+
+        private readonly IDogStatsd _statsd;
+        private readonly int _processId;
+
+        private int? _previousGen0Count;
+        private int? _previousGen1Count;
+        private int? _previousGen2Count;
+
+        private double? _lastContentionCount;
+        private MemoryMappedFile _file;
+        private MemoryMappedViewAccessor _view;
+
+        public MemoryMappedCounters(IDogStatsd statsd)
+        {
+            _statsd = statsd;
+
+            ProcessHelpers.GetCurrentProcessInformation(out _, out _, out _processId);
+
+            MemoryMappedFile file = null;
+            MemoryMappedViewAccessor view = null;
+
+            try
+            {
+                file = MemoryMappedFile.OpenExisting(@"Cor_CLR_WRITER\Cor_SxSPublic_IPCBlock");
+                view = file.CreateViewAccessor();
+
+                if (GC.CollectionCount(0) == 0)
+                {
+                    // This is unlikely to happen, but we must make sure that the GC has run at least once,
+                    // otherwise the GC structure that we use for the sanity check will be empty.
+                    GC.Collect(0, GCCollectionMode.Forced, blocking: true); // Sorry ._.
+                }
+
+                // Sanity check
+                view.Read<IpcControlBlock>(0, out var controlBlock);
+
+                if (controlBlock.Header.Flags == 0)
+                {
+                    throw new InvalidOperationException("The IPC control block is not initialized");
+                }
```

Comment:
Do we need to check if `IpcHeaderFlags.IPC_FLAG_INITIALIZED` is initialized

Reply:
Right, better safe than sorry